plugins/ua_nodestore_binary_encoded.c

```diff
@@ -0,0 +1,1163 @@
+        break;
+    case UA_NODECLASS_VARIABLETYPE:
+        nodeSize += variableTypeNodeCalcSizeBinary((const UA_VariableTypeNode*)node);
+        break;
+    case UA_NODECLASS_REFERENCETYPE:
+        nodeSize += referenceTypeNodeCalcSizeBinary((const UA_ReferenceTypeNode*)node);
+        break;
+    case UA_NODECLASS_DATATYPE:
+        nodeSize += dataTypeNodeCalcSizeBinary((const UA_DataTypeNode*)node);
+        break;
+    case UA_NODECLASS_VIEW:
+        nodeSize += viewNodeCalcSizeBinary((const UA_ViewNode*)node);
+        break;
+    default:
+        break;
+    }
+
+    return nodeSize;
+}
+
+static UA_StatusCode
+getEncodeNodes (const UA_Node *node, UA_ByteString *new_valueEncoding) {
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+
+    UA_Byte *bufPos = new_valueEncoding->data;
+    const UA_Byte *bufEnd = &new_valueEncoding->data[new_valueEncoding->length];
+    retval |= UA_NodeId_encodeBinary(&node->nodeId, &bufPos, bufEnd);
+    retval |= UA_NodeClass_encodeBinary(&node->nodeClass, &bufPos, bufEnd);
+    retval |= UA_QualifiedName_encodeBinary(&node->browseName, &bufPos, bufEnd);
+    retval |= UA_LocalizedText_encodeBinary(&node->displayName, &bufPos, bufEnd);
+    retval |= UA_LocalizedText_encodeBinary(&node->description, &bufPos, bufEnd);
+    retval |= UA_UInt32_encodeBinary(&node->writeMask, &bufPos, bufEnd);
+    retval |= UA_UInt64_encodeBinary((const UA_UInt64 *)&node->referencesSize, &bufPos, bufEnd);
+    for (size_t i = 0; i < node->referencesSize; i++) {
+        retval |= UA_NodeReferenceKind_encodeBinary(&node->references[i], &bufPos, &bufEnd);
+    }
+
+    switch(node->nodeClass) {
+    case UA_NODECLASS_OBJECT:
+        retval |= objectNodeEncode((const UA_ObjectNode*)node, bufPos, bufEnd);
+        break;
+    case UA_NODECLASS_VARIABLE:
+        retval |= variableNodeEncode((const UA_VariableNode*)node, bufPos, bufEnd);
+        break;
+    case UA_NODECLASS_METHOD:
+        retval |= methodNodeEncode((const UA_MethodNode*)node, bufPos, bufEnd);
+        break;
+    case UA_NODECLASS_OBJECTTYPE:
+        retval |= objectTypeNodeEncode((const UA_ObjectTypeNode*)node, bufPos, bufEnd);
+        break;
+    case UA_NODECLASS_VARIABLETYPE:
+        retval |= variableTypeNodeEncode((const UA_VariableTypeNode*)node, bufPos, bufEnd);
+        break;
+    case UA_NODECLASS_REFERENCETYPE:
+        retval |= ReferenceTypeNodeEncode((const UA_ReferenceTypeNode*)node, bufPos, bufEnd);
+        break;
+    case UA_NODECLASS_DATATYPE:
+        retval |= dataTypeNodeEncode((const UA_DataTypeNode*)node, bufPos, bufEnd);
+        break;
+    case UA_NODECLASS_VIEW:
+        retval |= viewNodeEncode((const UA_ViewNode*)node, bufPos, bufEnd);
+        break;
+    default:
+        break;
+    }
+
+    return retval;
+}
+#endif
+
+#ifdef UA_ENABLE_USE_ENCODED_NODES
+#define MINIMALNODECOUNT        45 // Total number of nodes in MINIMAL configuration
```

Comment:
I'm not sure why these 45 nodes get special treatment.
Can you add a bit of documentation?

I'd prefer (by far!) if we could handle all nodes the same for the binary nodestore.
Maybe we can just rip out the ns0 initialization in ua_server_ns0 if the binary nodestore is provided.
So we can assume they are "pre-initialized" in the binary nodestore.

Reply:
By the way, this contradicts my previous position that no changes in /src are permitted. 😄 
But changing one line (init_ns0) is less of a problem than tight integration.